### PR TITLE
Linking activities

### LINKED ISSUE #47: As a User I should be able to link existing activity
## I'm submitting a...

- [] Bug report  
- [ ] Regression (a behavior that used to work and stopped working in a new release)
- [x] Feature request
- [ ] Documentation issue or request

## Expected Behavior
In addition to creation and cloning, User should be able to link activity.

### PR DESCRIPTION
**Pull request by zcuric**
### This PR
- [x] Adds ability to link an activity 
- [x] Adds ability to clone linked activity
- [x]  Adds ability to publish linked activity

### Several notes for developers and QA
- When one activity is being linked, two new activities are created. Both of those activities are links, one will replace the original activity, another will be linked in desired place. The original activity becomes an `origin` and it is not represent in the UI. Each link contains `originId`. 
- Activity can be linked any where the same repository/course if it satisfies the conditions which are the same as cloning. 
- Activity can be linked or created inside another linked parent. In that case, all siblings of the linked parent will receive link and if link is delete for one linked parent it will deleted for all linked parents. 
- When updating a link, updating will forward to origin and origin will be updated, but all links will receive the update. 
- When removing linked activity, linked activity will be deleted, not detached and if the number of links is 1, the origin activity will be restored to place of last remaining link. 
- Only origins are published. Published activities in the collection receive `children` array with all of children `ids` and children can have  array of `parentIds`. So children can have multiple parents and for LMS it's to represent links. 
- Cloning is also enabled. It works basically the same as normal cloning. If whole repository is cloned, links will be cloned as well. If individual nodes are cloned, links will be cloned. If there is only one link in the node, it will be resolved to origin. 
- Cloning into link is also posible.

Closes #47 #46

### CODE REVIEW

**Comment by zcuric** [comment]
Blocked by #312 . Need a detail discussion on how to handle multiple additions into linked trees with @kronicker .

File: client/components/course/Outline/InsertActivity/SelectAction.vue [hunk 1]
```diff
@@ -16,11 +16,17 @@ export default {
...
         { label: 'Create', icon: 'plus', action: select('create') },
         { label: 'Copy', icon: 'content-copy', action: select('copy') },
         { label: 'Cancel', icon: 'close', action: () => this.$emit('close') }
       ];
+
+      if (!process.env.ENABLE_ACTIVITY_LINKING) return actions;
+      actions.splice(
+        2, 0, { label: 'Use existing', icon: 'link', action: select('link') },
+      );
+      return actions;
```
Comment (vladimyr) [inline client/components/course/Outline/InsertActivity/SelectAction.vue]: Hmm, you can make it more compact relying on object truthiness:
```js
const isLinkingSupported = process.env.ENABLE_ACTIVITY_LINKING;

const actions = [
  { label: 'Create', icon: 'plus', action: select('create') },
  { label: 'Copy', icon: 'content-copy', action: select('copy') },
  isLinkingSupported && 
    { label: 'Use existing', icon: 'link', action: select('link') },
  { label: 'Cancel', icon: 'close', action: () => this.$emit('close') }
].filter(Boolean);
```

Comment (zcuric) [reply client/components/course/Outline/InsertActivity/SelectAction.vue]: Me can do that. 🐒

File: client/components/common/ActivityBrowser/SelectActivity.vue [hunk 2]
```diff
@@ -88,7 +92,9 @@ export default {
     },
     getChildren(activity) {
       const parentId = get(activity, 'id', null);
-      return filter(this.activities, { parentId });
+      return filter(this.activities, it => {
```
Comment (abasic) [inline client/components/common/ActivityBrowser/SelectActivity.vue]:
```suggestion
      return filter(this.activities, { parentId, isOrigin: false });
```

File: client/components/common/ActivityBrowser/SelectActivity.vue [hunk 3]
```diff
@@ -64,6 +65,9 @@ export default {
       if (!this.selectableLevels.length) return true;
       const schema = getSchemaId(first(this.selectableLevels).type);
       return schema === this.repository.schema;
+    },
+    showBackButton() {
+      return !(this.action === 'link' && !this.parent);
```
Comment (abasic) [inline client/components/common/ActivityBrowser/SelectActivity.vue]:
```suggestion
      return this.action !== 'link' || this.parent;
```

File: client/components/course/Outline/Activity.vue [hunk 4]
```diff
@@ -19,13 +19,16 @@
...
+          <v-icon v-if="originId" class="pr-2 link-icon">mdi-link</v-icon> {{ data.name }}
         </span>
         <div v-show="isHighlighted" class="actions">
           <v-spacer />
           <v-btn
             v-show="isEditable"
-            :to="{ name: 'editor', params: { activityId: id } }"
+            :to="{
+              name: 'editor',
+              params: { activityId: originId ? originId : id }
```
Comment (abasic) [inline client/components/course/Outline/Activity.vue]:
```suggestion
              params: { activityId: originId || id }
```

File: client/components/course/Outline/Activity.vue [hunk 5]
```diff
@@ -212,4 +217,16 @@ export default {
 .sub-activity {
   margin-left: 44px;
 }
+
+.link {
+  background: #CCCC;
```
Comment (abasic) [inline client/components/course/Outline/Activity.vue]:
```suggestion
  background: #CCC;
```

File: client/components/course/Outline/InsertActivity/SelectAction.vue [hunk 6]
```diff
@@ -12,15 +12,18 @@
 </template>
 
 <script>
+import { mapGetters } from 'vuex';
```
Comment (abasic) [inline client/components/course/Outline/InsertActivity/SelectAction.vue]: Missing empty line after.

File: client/components/course/TreeView/TreeGraph.vue [hunk 7]
```diff
@@ -124,10 +124,12 @@ export default {
         .attr('class', d => `node depth-${d.depth}`)
         .attr('transform', d => `translate(${d.x}, ${d.y})`);
 
+      const label = ({ data }) => data.name || (data.originId || data.id);
```
Comment (abasic) [inline client/components/course/TreeView/TreeGraph.vue]:
```suggestion
      const label = ({ data }) => data.name || data.originId || data.id;
```

File: client/components/course/TreeView/TreeGraph.vue [hunk 8]
```diff
@@ -124,10 +124,12 @@ export default {
...
+
       // Append label.
       node.append('text')
         .classed('label', true)
-        .text(d => d.data.name || d.data.id)
+        .text(d => label(d))
```
Comment (abasic) [inline client/components/course/TreeView/TreeGraph.vue]:
```suggestion
        .text(label)
```

File: client/components/course/TreeView/index.vue [hunk 9]
```diff
@@ -87,8 +87,14 @@ function tree(activities, structure, root = { size: 0 }, parent = root, depth =
   parent.children = reduce(activities, (acc, it) => {
     const parentId = parent.id || null;
     if (it.parentId !== parentId) return acc;
+    if (it.parentId === parent.originId) {
```
Comment (abasic) [inline client/components/course/TreeView/index.vue]: Might be useless, check the use case.

File: client/components/course/Sidebar/Body.vue [hunk 10]
```diff
@@ -43,7 +43,7 @@ export default {
     ...mapActions('activities', ['update']),
     updateActivity(key, value) {
       const data = { ...this.activity.data, [key]: value };
-      this.update({ _cid: this.activity._cid, data });
+      this.update({ activity: this.activity, data });
```
Comment (abasic) [inline client/components/course/Sidebar/Body.vue]: Pass only `originId` and `_cid`, cause they are only used in `update` action.

File: server/activity/activity.controller.js [hunk 11]
```diff
@@ -77,14 +107,28 @@ function getPreviewUrl({ course, activity }, res) {
...
+const linkCreated = async activity => {
+  if (!activity.parentId) return activity;
+  const parent = await activity.getParent();
+  if (!parent.isLink) return activity;
+  await activity.link({
+    position: activity.position,
+    parentId: activity.parentId
+  });
+  await activity.update({ parentId: parent.origin.id });
+  return Activity.findByPk(activity.id).then(data => data);
```
Comment (abasic) [inline server/activity/activity.controller.js]:
```suggestion
  return Activity.findByPk(activity.id);
```

File: server/activity/activity.controller.js [hunk 12]
```diff
@@ -18,7 +21,9 @@ function create({ course, body, params, user }, res) {
     { data: Object.assign({}, defaultMeta, body.data) },
     { courseId: params.courseId });
   const opts = { context: { userId: user.id } };
-  return Activity.create(data, opts).then(data => res.json({ data }));
+  return Activity.create(data, opts)
```
Comment (abasic) [inline server/activity/activity.controller.js]: This should be inside a single transaction.

File: server/activity/activity.model.js [hunk 13]
```diff
@@ -96,31 +120,165 @@ class Activity extends Model {
...
-      const where = { activityId: it.id, detached: false };
-      const tes = await TeachingElement.findAll({ where, transaction });
-      await TeachingElement.cloneElements(tes, parent, transaction);
-      const children = await it.getChildren({ where: { detached: false } });
+  get isLink() {
+    return !!(this.originId && this.origin);
+  }
+
+  get isOrigin() {
+    return this.links && this.links.length > 0;
```
Comment (abasic) [inline server/activity/activity.model.js]:
```suggestion
    return this.links && this.links.length;
```

File: server/activity/activity.model.js [hunk 14]
```diff
@@ -201,11 +424,195 @@ class Activity extends Model {
...
+ * @param {Object} options
+ * @param {Number} options.courseId
+ * @param {Number} options.parentId
+ * @param {Object} options.idMap
+ * @param {Boolean} options.cloneOrigins
+ */
+function getClonedActivitiesMap(source, options) {
+  const { idMap, cloneOrigins, courseId, parentId } = options;
+  const originId = ({ originId }) =>
+    cloneOrigins && idMap[originId] ? idMap[originId] : originId;
```
Comment (abasic) [inline server/activity/activity.model.js]:
```suggestion
    cloneOrigins && (idMap[originId] || originId);
```

File: server/activity/activity.model.js [hunk 15]
```diff
@@ -96,31 +120,165 @@ class Activity extends Model {
...
+    if (!isLinkingEnabled) return idMap;
+    const activities = await Activity.findAll({ where: { courseId } });
+    const origins = activities.filter(({ isOrigin }) => isOrigin);
+    if (!origins.length) return idMap;
+    await Promise.each(origins, async origin => {
+      const links = await origin.getLinks({ transaction });
+      if (links.length !== 1) return;
+      const [link] = links;
+      await resolveOriginWithOneLink(link, { transaction });
+      idMap = omitBy(idMap, id => id === link.id);
```
Comment (abasic) [inline server/activity/activity.model.js]:
```suggestion
      idMap = omitBy(idMap, { id: link.id });
```

File: server/activity/activity.model.js [hunk 16]
```diff
@@ -96,31 +120,165 @@ class Activity extends Model {
...
+  /**
+   * Link activity
+   * @param {Object} options
+   * @param {Number} options.parentId
+   * @param {Number} options.position
+   */
+  link({ parentId, position }) {
+    return this.sequelize.transaction(transaction => {
+      return Activity.linkActivities(this, { parentId, position, transaction })
+        .then(activities => activities);
```
Comment (abasic) [inline server/activity/activity.model.js]: No need for this line.

File: server/activity/activity.model.js [hunk 17]
```diff
@@ -167,30 +325,95 @@ class Activity extends Model {
...
+        ...descendants.nodes.filter(d => !d.isLink),
+        ...descendants.leaves.filter(d => !d.isLink)
+      ];
+      descendants.links = descendants.nodes.filter(d => d.isLink);
+      let where = { activityId: [...map(descendants.all, 'id'), this.id] };
+      await removeAll(TeachingElement, where, options);
+      where = { parentId: [...map(descendants.nodes.filter(d => !d.isLink), 'id'), this.id] };
+      await removeAll(Activity, where, options);
+      if (descendants.links.length) {
+        await Promise.each(descendants.links, async link => {
```
Comment (abasic) [inline server/activity/activity.model.js]: No need for `async` function inside `Promise.each`.

File: server/activity/activity.model.js [hunk 18]
```diff
@@ -201,11 +424,195 @@ class Activity extends Model {
...
+/**
+ * @param  {Activity} link
+ * @param  {Object} options
+ */
+async function resolveOriginWithOneLink(link, options) {
+  const children = await link.getChildren(options);
+  if (children.length) {
+    await Promise.each(
+      children,
+      async child => {
```
Comment (abasic) [inline server/activity/activity.model.js]: This does not need to be an `async` function.

File: server/activity/activity.model.js [hunk 19]
```diff
@@ -201,11 +424,195 @@ class Activity extends Model {
...
+/**
+ * @param {Activity} source
+ * @param {Number} parentId
+ * @returns {Boolean}
+ */
+function addSourceLink(source, parentId) {
+  if (!source.isLink && !source.isOrigin) return true;
+  if (!parentId && source.isLink) return false;
+  if (!parentId) return true;
+  if (source.parentId === parentId) return false;
```
Comment (abasic) [inline server/activity/activity.model.js]: This line is not necessary.

File: server/activity/activity.model.js [hunk 20]
```diff
@@ -201,11 +424,195 @@ class Activity extends Model {
...
+
+/**
+ * @param {Activity} source
+ * @param {Number} parentId
+ * @returns {Boolean}
+ */
+function addSourceLink(source, parentId) {
+  if (!source.isLink && !source.isOrigin) return true;
+  if (!parentId && source.isLink) return false;
+  if (!parentId) return true;
```
Comment (abasic) [inline server/activity/activity.model.js]: You can just return `!parentId` here.

File: server/course/course.model.js [hunk 21]
```diff
@@ -67,6 +73,10 @@ class Course extends Model {
     });
   }
 
+  get isLinkingEnabled() {
```
Comment (abasic) [inline server/course/course.model.js]: This and the virtual field are the same. Only one needs to be kept.

File: server/shared/database/migrations/20190605084444-position-can-be-null.js [hunk 22]
```diff
@@ -0,0 +1,20 @@
+'use strict';
```
Comment (abasic) [inline server/shared/database/migrations/20190605084444-position-can-be-null.js]: No need for this migration.

File: server/shared/publishing/helpers.js [hunk 23]
```diff
@@ -41,6 +43,26 @@ function publishActivity(activity) {
     });
   });
 }
+async function publishLinkedActivity(activity) {
```
Comment (abasic) [inline server/shared/publishing/helpers.js]: Missing empty line above.

File: server/shared/publishing/helpers.js [hunk 24]
```diff
@@ -41,6 +43,26 @@ function publishActivity(activity) {
...
+  return getStructureData(activity).then(async data => {
+    let { repository, predecessors, spine } = data;
+    predecessors.forEach(async it => {
+      const exists = find(spine.structure, { id: it.id });
+      if (!exists) {
```
Comment (abasic) [inline server/shared/publishing/helpers.js]: You can early return here.

File: server/shared/publishing/helpers.js [hunk 25]
```diff
@@ -41,6 +43,26 @@ function publishActivity(activity) {
...
+        if (it.isLink) await addLinkToSpine(spine, it);
+        else await addToSpine(spine, it, repository);
```
Comment (abasic) [inline server/shared/publishing/helpers.js]: This can use a ternary operator here.

File: server/shared/publishing/helpers.js [hunk 26]
```diff
@@ -224,33 +293,103 @@ function saveSpine(spine) {
...
   );
-  renameKey(activity, 'data', 'meta');
-  let index = findIndex(spine.structure, { id: activity.id });
+
+  if (isLinkingEnabled) {
+    const children = await activity.getChildren();
+    let childrenIds = [];
+    if (children.length) {
+      childrenIds = uniq(children.sort((a, b) => a.position > b.position).map(child => {
+        return child.originId ? child.originId : child.id;
```
Comment (abasic) [inline server/shared/publishing/helpers.js]:
```suggestion
        return child.originId || child.id;
```

File: server/shared/publishing/helpers.js [hunk 27]
```diff
@@ -224,33 +293,103 @@ function saveSpine(spine) {
...
     }
   );
-  renameKey(activity, 'data', 'meta');
-  let index = findIndex(spine.structure, { id: activity.id });
+
+  if (isLinkingEnabled) {
+    const children = await activity.getChildren();
+    let childrenIds = [];
+    if (children.length) {
+      childrenIds = uniq(children.sort((a, b) => a.position > b.position).map(child => {
```
Comment (abasic) [inline server/shared/publishing/helpers.js]: Sort children in the line above for the purpose of more readable code.

File: server/shared/publishing/helpers.js [hunk 28]
```diff
@@ -224,33 +293,103 @@ function saveSpine(spine) {
...
+          if (it.id === spineParent.id) return spineParent;
+          return it;
+        });
+      }
+    }
+  }
+
+  if (index > -1) {
+    const parentIds = Array.isArray(spine.structure[index].parentId)
+      ? spine.structure[index].parentId : [spine.structure[index].parentId];
```
Comment (abasic) [inline server/shared/publishing/helpers.js]: This should be two lines.

File: server/shared/publishing/helpers.js [hunk 29]
```diff
@@ -224,33 +293,103 @@ function saveSpine(spine) {
...
+    return;
+  }
+
+  const children = await activity.getChildren();
+  let childrenIds = [];
+  if (children.length) {
+    childrenIds = uniq(
+      children.sort((a, b) => a.position > b.position)
+      .map(child => {
+        return child.originId ? child.originId : child.id;
```
Comment (abasic) [inline server/shared/publishing/helpers.js]:
```suggestion
        return child.originId || child.id;
```

**Comment by underscope** [comment]
Per our recent discussion. Closing this one until we do BE cleanup and implement graph.

**Pull request closed without merging**